File: week1/credit.c
#include <cs50.h>
#include <stdio.h>

int check_Luhn(long number);
int get_digits(long number);
int digit_sum(int x);
int get_start_number(long number, int digits_want);

int main(void)
{
    int valid = 0;
    long card_number = get_long("Number: ");
    if (card_number > 0 && check_Luhn(card_number))
    {
        valid = 1;
    }
    else
    {
        valid = 0;
    }

    if (valid)
    {
        int digits = get_digits(card_number);
        if (digits == 15 && 
            (get_start_number(card_number, 2) == 34 || get_start_number(card_number, 2) == 37))
        {
            printf("AMEX\n");
        }
        else if (digits == 16 && 
                 (51 <= get_start_number(card_number, 2) && get_start_number(card_number, 2) <= 55))
        {
            printf("MASTERCARD\n");
        }
        else if ((digits == 13 || digits == 16) && get_start_number(card_number, 1) == 4)
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }

    return 0;
}

int digit_sum(int x)
{
    int ret = 0;
    while (x > 0)
    {
        ret += x % 10;
        x /= 10;
    }

    return ret;
}

int check_Luhn(long number)
{
    int check_sum = 0;

    long tmp = number / 10;
    while (tmp > 0)
    {
        check_sum += digit_sum((tmp % 10) * 2);
        tmp /= 100;
    }

    tmp = number;
    while (tmp > 0)
    {
        check_sum += (tmp % 10);
        tmp /= 100;
    }

    return ((check_sum % 10) == 0);
}

int get_digits(long number)
{
    int ret = 0;
    while (number > 0)
    {
        number /= 10;
        ret++;
    }

    return ret;
}

int get_start_number(long number, int digits_want)
{
    while (number > 0)
    {
        if (get_digits(number) == digits_want)
        {
            return number;
        }
        number /= 10;
    }
    return number;
}

File: week6/credit.py
number = input("Number: ")


def luhn(number) -> bool:
    sum_val = 0
    sum_val += sum(list(map(lambda x: sum([int(s) for s in str(int(x)*2)]), number[-2::-2])))
    sum_val += sum(list(map(int, number[-1::-2])))
    print("sum: ", sum_val)
    return sum_val % 10 == 0


if luhn(number):
    if len(number) == 15 and number[:2] in ["34", "37"]:
        print("AMEX")
    elif len(number) == 16 and number[:2] in ["51", "52", "53", "54", "55"]:
        print("MASTERCARD")
    elif len(number) in [13, 16] and number[0] == "4":
        print("VISA")
    else:
        print("INVALID")
else:
    print("INVALID")
